homeassistant/components/netgear/config_flow.py

```diff
@@ -0,0 +1,179 @@
+        if not user_input:
+            user_input = {}
+
+        if self.discovered:
+            data_schema = _discovery_schema_with_defaults(user_input)
+        else:
+            data_schema = _user_schema_with_defaults(user_input)
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=data_schema,
+            errors=errors or {},
+            description_placeholders=self.placeholders,
+        )
+
+    async def async_step_import(self, user_input=None):
+        """Import a config entry."""
+        return await self.async_step_user(user_input)
+
+    async def async_step_ssdp(self, discovery_info: dict) -> FlowResult:
+        """Initialize flow from ssdp."""
+        updated_data = {}
+
+        device_url = urlparse(discovery_info[ssdp.ATTR_SSDP_LOCATION])
+        if device_url.hostname:
+            updated_data[CONF_HOST] = device_url.hostname
+        if device_url.port:
+            updated_data[CONF_PORT] = device_url.port
+        if device_url.scheme == "https":
+            updated_data[CONF_SSL] = True
+        else:
+            updated_data[CONF_SSL] = False
+
+        await self.async_set_unique_id(discovery_info[ssdp.ATTR_UPNP_SERIAL])
+        self._abort_if_unique_id_configured(updates=updated_data)
+        self.placeholders.update(updated_data)
+        self.discovered = True
+
+        return await self.async_step_user()
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initiated by the user."""
+        errors = {}
+
+        if user_input is None:
+            return await self._show_setup_form()
+
+        host = user_input.get(CONF_HOST, self.placeholders[CONF_HOST])
+        port = user_input.get(CONF_PORT, self.placeholders[CONF_PORT])
+        ssl = user_input.get(CONF_SSL, self.placeholders[CONF_SSL])
+        username = user_input.get(CONF_USERNAME, self.placeholders[CONF_USERNAME])
+        password = user_input[CONF_PASSWORD]
+        if not username:
+            username = self.placeholders[CONF_USERNAME]
+
+        # Open connection and check authentication
+        try:
+            api = await self.hass.async_add_executor_job(
+                get_api, password, host, username, port, ssl
+            )
+        except CannotLoginException:
+            errors["base"] = "config"
+
+        if errors:
+            return await self._show_setup_form(user_input, errors)
+
+        # Check if already configured
+        info = await self.hass.async_add_executor_job(api.get_info)
+        await self.async_set_unique_id(info["SerialNumber"], raise_on_progress=False)
+        self._abort_if_unique_id_configured()
+
+        config_data = {
+            CONF_USERNAME: username,
+            CONF_PASSWORD: password,
+            CONF_HOST: host,
+            CONF_PORT: port,
+            CONF_SSL: ssl,
+        }
+
+        return self.async_create_entry(
+            title=f"{info['ModelName']} - {info['DeviceName']}",
```

Comment:
I'm getting this error while testing this PR:
```
2021-09-13 12:18:01 DEBUG (SyncWorker_7) [pynetgear] Login v1
2021-09-13 12:18:01 DEBUG (SyncWorker_0) [pynetgear] Get Info
2021-09-13 12:18:01 ERROR (MainThread) [aiohttp.server] Error handling request
Traceback (most recent call last):
  File "/home/martin/.virtualenvs/home-assistant/lib/python3.8/site-packages/aiohttp/web_protocol.py", line 422, in _handle_request
    resp = await self._request_handler(request)
  File "/home/martin/.virtualenvs/home-assistant/lib/python3.8/site-packages/aiohttp/web_app.py", line 499, in _handle
    resp = await handler(request)
  File "/home/martin/.virtualenvs/home-assistant/lib/python3.8/site-packages/aiohttp/web_middlewares.py", line 119, in impl
    return await handler(request)
  File "/home/martin/dev/home-assistant/core/homeassistant/components/http/security_filter.py", line 60, in security_filter_middleware
    return await handler(request)
  File "/home/martin/dev/home-assistant/core/homeassistant/components/http/forwarded.py", line 89, in forwarded_middleware
    return await handler(request)
  File "/home/martin/dev/home-assistant/core/homeassistant/components/http/request_context.py", line 24, in request_context_middleware
    return await handler(request)
  File "/home/martin/dev/home-assistant/core/homeassistant/components/http/ban.py", line 78, in ban_middleware
    return await handler(request)
  File "/home/martin/dev/home-assistant/core/homeassistant/components/http/auth.py", line 144, in auth_middleware
    return await handler(request)
  File "/home/martin/dev/home-assistant/core/homeassistant/components/http/view.py", line 135, in handle
    result = await result
  File "/home/martin/dev/home-assistant/core/homeassistant/components/config/config_entries.py", line 155, in post
    return await super().post(request, flow_id)
  File "/home/martin/dev/home-assistant/core/homeassistant/components/http/data_validator.py", line 63, in wrapper
    result = await method(view, request, *args, **kwargs)
  File "/home/martin/dev/home-assistant/core/homeassistant/helpers/data_entry_flow.py", line 109, in post
    result = await self._flow_mgr.async_configure(flow_id, data)
  File "/home/martin/dev/home-assistant/core/homeassistant/data_entry_flow.py", line 202, in async_configure
    result = await self._async_handle_step(flow, cur_step["step_id"], user_input)
  File "/home/martin/dev/home-assistant/core/homeassistant/data_entry_flow.py", line 260, in _async_handle_step
    result: FlowResult = await getattr(flow, method)(user_input)
  File "/home/martin/dev/home-assistant/core/homeassistant/components/netgear/config_flow.py", line 177, in async_step_user
    title=f"{info['ModelName']} - {info['DeviceName']}",
KeyError: 'DeviceName'
```
We probably need to adapt to which router category we're interfacing with.

Reply:
I just implemented a fix in the latest commit, could you test again?

I tried again with my Netgear R7000, but for me it works fine and the DeviceName is available